sys/net/application_layer/gcoap/forward_proxy.c

```diff
@@ -50,8 +54,61 @@ gcoap_listener_t forward_proxy_listener = {
 void gcoap_forward_proxy_init(void)
 {
     gcoap_register_listener(&forward_proxy_listener);
+
+    /* initialize the nanocoap cache operation, if compiled */
+    if (IS_ACTIVE(MODULE_NANOCOAP_CACHE)) {
+        nanocoap_cache_init();
+    }
+}
+
+static int _cache_build_response(nanocoap_cache_entry_t *ce,
+                                 coap_pkt_t *pdu,
+                                 uint8_t *buf,
+                                 size_t len)
+{
+    /* Use the same code from the cached content. Use other header
+     * fields from the incoming request */
+    gcoap_resp_init(pdu, buf, len, ce->response_pkt.hdr->code);
+    /* copy all options and possible payload from the cached response
+     * to the new response */
+    unsigned header_len_req = coap_get_total_hdr_len(pdu);
+    unsigned header_len_cached = coap_get_total_hdr_len(&ce->response_pkt);
+    unsigned opt_payload_len = ce->response_len - header_len_cached;
+
+    memcpy((buf + header_len_req),
+           (ce->response_buf + header_len_cached),
+           opt_payload_len);
+    return header_len_req + opt_payload_len;
+}
+
+static int _cache_lookup_and_process(coap_pkt_t *pdu,
+                                     uint8_t *buf,
+                                     size_t len,
+                                     client_ep_t *cep)
+{
+    (void) cep;
+
+    uint8_t cache_key[SHA256_DIGEST_LENGTH];
+    ztimer_now_t now = ztimer_now(ZTIMER_SEC);
+    nanocoap_cache_key_generate(pdu, cache_key);
+    nanocoap_cache_entry_t *ce = nanocoap_cache_key_lookup(cache_key);
+
+    /* cache hit, methods are equal, and cache entry is not stale */
+    if (ce &&
+        (ce->request_method == coap_get_code(pdu)) &&
+        (ce->max_age > now)) {
+        /* use response from cache */
+        return _cache_build_response(ce, pdu, buf, len);
+    }
+
+#if IS_ACTIVE(MODULE_NANOCOAP_CACHE)
```

Comment:
Here and elsewhere, where `IS_ACTIVE(MODULE_*)` is used:

```suggestion
#if IS_USED(MODULE_NANOCOAP_CACHE)
```